### PR TITLE
Vivax remove severe

### PR DESCRIPTION
**Pull request by RJSheppard**
Added parasite specific immunities (falciparum: ib, id, ica, icm, iva, ivm; vivax: id, idm, ica, icm). These affect probability of disease and probability of patent (LM-detectable) infections, with initiation at birth/death, decay functions and boosts. id vivax also impacts duration of sub-patent infection (previously coded). Could rename id as ida for consistency with ica/icm?

### CODE REVIEW

File: R/human_infection.R [hunk 1]
```diff
@@ -400,14 +409,44 @@ schedule_infections <- function(
...
-    update_to_asymptomatic_infection(
-      variables,
-      parameters,
-      timestep,
-      to_infect_asym
-    )
+  
+  # falciparum infection can result in D or A
+  if(parameters$parasite == "falciparum"){
+    to_infect_asym <- clinical_infections$copy()$not(FALSE)$and(infections)$and(included)
```
Comment (giovannic) [inline R/human_infection.R:415]: If $not is returning a copy you do not need to copy beforehand: 

```suggestion
    to_infect_asym <- clinical_infections$not(FALSE)$and(infections)$and(included)
```

File: R/human_infection.R [hunk 2]
```diff
@@ -400,14 +409,44 @@ schedule_infections <- function(
...
+      update_to_asymptomatic_infection(
+        variables,
+        parameters,
+        timestep,
+        to_infect_asym
+      )}
+    
+    # vivax infection can result in D, A or U  
+  } else if (parameters$parasite == "vivax"){
+    to_infect_asym <- patent_infections$copy()$and(included)$and(clinical_infections$not(FALSE))
```
Comment (giovannic) [inline R/human_infection.R:428]:
```suggestion
    to_infect_asym <- patent_infections$copy()$and(included)$and(clinical_infections$not(TRUE))
```

File: R/human_infection.R [hunk 3]
```diff
@@ -400,14 +409,44 @@ schedule_infections <- function(
...
+    to_infect_subpatent <- patent_infections$copy()$not(FALSE)$and(included)
```
Comment (giovannic) [inline R/human_infection.R:429]:
```suggestion
    to_infect_subpatent <- patent_infections$not(TRUE)$and(included)
```

File: R/human_infection.R [hunk 4]
```diff
@@ -371,25 +381,24 @@ calculate_treated <- function(
...
   variables,
+  patent_infections = NULL,
   clinical_infections,
   treated,
   infections,
   parameters,
   timestep
   ) {
-  included <- treated$not(TRUE)
+  included <- treated$not(FALSE)
```
Comment (giovannic) [inline R/human_infection.R:399]:
```suggestion
  included <- treated$not(TRUE)
```

File: R/human_infection.R [hunk 5]
```diff
@@ -371,25 +381,24 @@ calculate_treated <- function(
 #' @description
 #' Schedule infections in humans after the incubation period
 #' @param events a list of all of the model events
+#' @param patent_infections bitset of patent-level infected humans (P.v only)
 #' @param clinical_infections bitset of clinically infected humans
 #' @param treated bitset of treated humans
 #' @param infections bitset of infected humans
 #' @param parameters model parameters
 #' @noRd
 schedule_infections <- function(
```
Comment (giovannic) [inline R/human_infection.R:390]: We don't use any of these variables again after schedule_infections right? Let's avoid copying them loads of times.

Hopefully, your test will confirm that this works.

**Review by giovannic — CHANGES_REQUESTED** [review]
Great stuff!

File: R/variables.R [hunk 6]
```diff
@@ -235,7 +243,15 @@ create_variables <- function(parameters) {
     spray_time = spray_time
   )
   
-  if(parameters$parasite == "vivax"){
+  if(parameters$parasite == "falciparum"){
```
Comment (pwinskill) [inline R/variables.R:246]:
```suggestion
  if(parameters$parasite == "falciparum"){
    variables <- c(
      variables, 
      last_boosted_ib = last_boosted_ib,
      last_boosted_iva = last_boosted_iva,
      ivm = ivm,
      ib = ib,
      iva = iva
    )
  } else if(parameters$parasite == "vivax"){
    variables <- c(
      variables, 
      idm = idm
    )
  }
```

File: tests/testthat/test-vivax.R [hunk 7]
```diff
@@ -174,23 +176,123 @@ test_that('that vivax patent incidence rendering works', {
...
+
+  # falciparum asymptomatic infectivity should not equal ca
+  expect_false(any(
+    c(create_variables(parameters)$infectivity$get_values(),
+      create_variables(dem_parameters)$infectivity$get_values())==vivax_parameters$ca))
+
+  # falciparum asymptomatic infectivity should change with age structure
+  expect_false(any(
+    c(create_variables(parameters)$infectivity$get_values() == create_variables(dem_parameters)$infectivity$get_values())))
+})
```
Comment (pwinskill) [inline tests/testthat/test-vivax.R:254]: I'm wondering if these checks for things not matching here are a little fragile? presumably with some set of reasonable parameters one or more of these values could `==ca` or could be the same with or without the age structure?

Would it be possible to simplify by
1. Initialising a population with 1 person at a given age
2. Running create variables, checking that the infectivity values are = to expected for both falicaparum and vivax
3. Change the age and repeat 2?

**Review by pwinskill — CHANGES_REQUESTED** [review]
💫

File: R/human_infection.R [hunk 8]
```diff
@@ -128,8 +133,13 @@ calculate_infections <- function(
   source_humans <- variables$state$get_index_of(
     c('S', 'A', 'U'))$and(bitten_humans)
 
-  b <- blood_immunity(variables$ib$get_values(source_humans), parameters)
-
+  if(parameters$parasite == "falciparum"){ ## P. falciparum models blood immunity
+    b <- blood_immunity(variables$ib$get_values(source_humans), parameters)
+    
+  } else if (parameters$parasite == "vivax"){ ## P. vivax does not model blood immunity
```
[the remark this answers is not in the document]
Comment (RJSheppard) [reply R/human_infection.R:139]: I believe blood immunity decreases probability of infection. This doesn't get modelled at all in vivax, as far as I've been able to work out. P. vivax has antiparasite immunity that determines patent from subpatent infections and decreases duration of sub-patent infection, and clinical immunity.

File: R/variables.R (hunk 6 above)
Comment (RJSheppard) [reply R/variables.R:246]: Corrected

File: tests/testthat/test-vivax.R [hunk 9]
```diff
@@ -174,23 +176,123 @@ test_that('that vivax patent incidence rendering works', {
...
+                                                t_proportion = 0,
+                                                init_id  = 0.5))
+  
+  vivax_parameters <- get_parameters(parasite = "vivax",
+                                     overrides = list(s_proportion = 0,
+                                                      d_proportion = 0,
+                                                      a_proportion = 1,
+                                                      u_proportion = 0,
+                                                      t_proportion = 0,
+                                                      init_id  = 0.5))
```
[the remark this answers is not in the document]
Comment (RJSheppard) [reply tests/testthat/test-vivax.R:218]: Corrected.

File: tests/testthat/test-vivax.R (hunk 7 above)
Comment (RJSheppard) [reply tests/testthat/test-vivax.R:254]: Changed this test.

File: R/human_infection.R (hunk 1 above)
Comment (RJSheppard) [reply R/human_infection.R]: I've played around with these a bit to make sure the outputs are being generated, and I think it's working now, including the corrections you suggested.

File: R/human_infection.R (hunk 2 above)
Comment (RJSheppard) [reply R/human_infection.R]: Changed

File: R/human_infection.R (hunk 4 above)
Comment (RJSheppard) [reply R/human_infection.R]: Changed

**Review by giovannic — APPROVED** [review]

**Pull request merged**